File: python/twitter.py
import heapq

class Twitter:

    def __init__(self):
        self.tweets = {}      # user_id: [(1, id), (4, id)]
        self.following = {}   # user_id: [following list]
        self.timestamp = 1

    def postTweet(self, userId: int, tweetId: int) -> None:
        if userId in self.tweets:
            self.tweets[userId].append((self.timestamp, tweetId))
        else:
            self.tweets[userId] = [(self.timestamp, tweetId)]
        self.timestamp += 1

    def getNewsFeed(self, userId: int) -> List[int]:
        heap = []
        users = self.following.get(userId, []).copy()
        users.append(userId)
        news_feed = []

        for user in users:
            if user in self.tweets and self.tweets[user]:
                idx = len(self.tweets[user]) - 1
                tweet_id = self.tweets[user][idx][1]
                heapq.heappush_max(heap, (self.tweets[user][idx][0], tweet_id, user, idx))
        while heap and len(news_feed) < 10:
            time, tweet_id, user_id, idx = heapq.heappop_max(heap)
            news_feed.append(tweet_id)

            if idx > 0:
                heapq.heappush_max(heap, (self.tweets[user_id][idx - 1][0], self.tweets[user_id][idx - 1][1], user_id, (idx - 1)))

        return news_feed

    def follow(self, followerId: int, followeeId: int) -> None:
        if followerId in self.following:
            self.following[followerId].append(followeeId)
        else:
            self.following[followerId] = [followeeId]

    def unfollow(self, followerId: int, followeeId: int) -> None:
        if followeeId in self.following[followerId]:
            self.following[followerId].remove(followeeId)

obj = Twitter()
obj.postTweet(1, 78)
obj.postTweet(2, 89)
obj.postTweet(3, 65)
obj.postTweet(1, 84)
obj.postTweet(3, 81)
obj.follow(1, 3)
print(obj.getNewsFeed(1))

File: cpp/twitter.cpp
#include <vector>
#include <unordered_map>
#include <queue>
using namespace std;

class Twitter {
    unordered_map<int, vector<int>> following;
    unordered_map<int, vector<pair<size_t, int>>> tweets;
    size_t timestamp;
public:
    Twitter() {
    	timestamp = 0;
    }

    void postTweet(int userId, int tweetId) {
        if (tweets.find(userId) == tweets.end()) {
            tweets.insert({userId, {}});
        }
	pair<size_t, int> tweet_data = {timestamp++, tweetId};
        tweets[userId].push_back(tweet_data);
    }

    vector<int> getNewsFeed(int userId) {
		vector<int> feed;
		priority_queue<pair<size_t, int>> pq;
		vector<int> following_list = following[userId];
		for (int user_id : following_list) {
			vector<pair<size_t, int>> user_tweets = tweets[user_id];
			for (int i = 0; i < user_tweets.size(); i++) {
				pq.push(user_tweets[i]);
			}
		}
		vector<pair<size_t, int>> own_tweets = tweets[userId];
		for (int i = 0; i < own_tweets.size(); i++) {
			pq.push(own_tweets[i]);
		}
		for (int i = 0; i < 10 && !pq.empty(); i++) {
			pair<size_t, int> topper = pq.top();
			pq.pop();
			feed.push_back(topper.second);
		}
		return feed;
    }

    void follow(int followerId, int followeeId) {
    	auto& f = following[followerId];
	if (find(f.begin(), f.end(), followeeId) == f.end()) {
		f.push_back(followeeId);
	}
    }

	void unfollow(int followerId, int followeeId) {
	    auto it = following.find(followerId);
	    if (it == following.end()) return;
	    auto& f = it->second;
	    auto pos = find(f.begin(), f.end(), followeeId);
	    if (pos != f.end()) f.erase(pos);
	    if (f.empty()) following.erase(it);
	}
};
